Денис Баутин

Duisus opened the pull request:
@P572

Commit:
Add new project TagsCloudVisualization

Commit:
Create class Spiral and tests for him

Add file Spiral.cs
Add file SpiralTests.cs

Commit:
Create extension for Rectangle and tests for him

Add file RectangleExtension.cs
Add file RectangleExtensionTests.cs

Commit:
Create class Surface and tests for him

Add file Surface.cs
Add file SurfaceTests.cs

Commit:
Create class CircularCloudLayouter and tests for him

Add file CircularCloudLayouter.cs
Add file CircularCloudLayouterTests.cs

Commit:
Create class Drawer

Add file Drawer.cs

Commit:
Modified tests for CircularCloudLayouter.
If test failed, image of layout will save.

Change file CircularCloudLayouterTests.cs

Commit:
Create examples of layout

Add file image1.png
Add file image2.png
Add file image3.png
Add file image4.png

Commit:
Create README.md with examples

Add file README.md

P572 commented on cs/TagsCloudVisualization/CircularCloudLayouter.cs:
```diff
@@ -0,0 +1,68 @@
+            foreach (var quarter in rectQuarters)
+            {
+                var (dx, dy) = GetDeltaForQuarter(quarter);
+                rect.Offset(dx, dy);
```
1) неочевидно, зачем нужно дополнительно сдвигать?
2) Хранение информации о четвертях(в т.ч. их сдвиги) звучит как ответственность Surface. То же самое относится и к сдвигам в сторону центра - если только поверхность обладает необходимой для сдвига прямоугольника к центру информацией, то она это и должна делать. Это можно сделать например через метод `AddRectangle `-  пусть на вход получает несдвинутый к центру прямоугольник, а возвращает уже сдвинутый, тот что и был добавлен.

P572 commented on cs/TagsCloudVisualization/CircularCloudLayouterTests.cs:
```diff
@@ -0,0 +1,90 @@
+
+        private static bool AreRectangleItersect(List<Rectangle> rects)
+        {
+            for (int i = 0; i < rects.Count - 1; i++)
```
Тоже можно упростить и избавиться от for

Commit:
Create test project and move the tests to it

Create TagsCloudVisualizationTests.csproj

Commit:
Replace RectangleExtension on RectangleDecorator, create tests for RectangleDecorator

Add file RectangleDecorator.cs
Add file RectangleDecoratorTests.cs
Delete file RectangleExtension.cs
Delete file RectangleExtensionTests.cs

Commit:
Reduced nesting in TearDown, refactor AreRectangleIntersect method

Change file CircularCloudLayouterTests.cs

Commit:
Rename method Surface.IsRectangleIntersect to Surface.RectangleIntersectsWithOther

Change files CircularCloudLayouter.cs, Surface.cs

Commit:
Move methods from CircularCloudLayouter to Surface (these methods move rectangle to center)
Add new tests for Surface

Change files CircularCloudLayouter.cs, Surface.cs, SurfaceTests.cs

Commit:
Create new examples of layout
Delete RectangleExtension

Change files image1.png, image2.png, image3.png, image4.png
Delete file RectangleExtension.cs

Commit:
Refactor Surface.GetResultDeltaPoint

Change file Surface.cs
